disaggregated_memory/boards/riscv_main_board.py

```diff
@@ -0,0 +1,596 @@
+        """
+        # get local memory is called at init phase.
+        return self._localMemory
+
+    def get_remote_memory(self) -> "AbstractMemorySystem":
+        """Get the memory (RAM) connected to the board.
+            This has to be implemeted by the child class as we don't know if
+            this board is simulating Gem5 memory or some external simulator
+            memory.
+        :returns: The remote memory system.
+        """
+        return self._remoteMemory
+
+    def get_remote_memory_size(self) -> "str":
+        """Get the remote memory size to setup the NUMA nodes. Since the remote
+            memory is an abstract memory system, we should be able to call its
+            standard methods.
+        :returns: The size of the remote memory system.
+        """
+        return self.get_remote_memory().get_size()
+
+    @overrides(RiscvBoard)
+    def get_mem_ports(self) -> Sequence[Tuple[AddrRange, Port]]:
+        return self.get_local_memory().get_mem_ports()
+
+    def get_remote_mem_ports(self) -> Sequence[Tuple[AddrRange, Port]]:
+        """Get the memory (RAM) ports connected to the board.
+            This has to be implemented by the child class as we don't know if
+            this board is simulating Gem5 memory or some external simulator
+            memory.
+        :returns: A tuple of mem_ports.
+        """
+        return self.get_remote_memory().get_mem_ports()
+
+    def get_remote_memory_addr_range(self):
+        """Get the range of the remote memory. This can be omitted in the
+            future iteration of the board.
+        :returns: AddrRange of the remote memory
+        """
+        # Although this is hardcoded to return the first element, this is
+        # always valid. This is how the standard library returns
+        # get_mem_ports().
+        if self._remoteMemoryAddressRange is None:
+            return self.get_remote_mem_ports()[0][0]
+        else:
+            return self._remoteMemoryAddressRange
+
+    @overrides(RiscvBoard)
+    def _setup_memory_ranges(self):
+        # the memory has to be setup for both the memory ranges. there is one
+        # local memory range, close to the host machine and the other range is
+        # pure memory, far from the host.
+        local_memory = self.get_local_memory()
+        # remote_memory = self.get_remote_memory_size()
+
+        local_mem_size = local_memory.get_size()
+        remote_mem_size = self.get_remote_memory_size()
+
+        # local memory range will always start from 0x80000000. The remote
+        # memory can start and end anywhere as long as it is consistent
+        # with the dtb.
+        self._local_mem_ranges = [
+            AddrRange(start=0x80000000, size=local_mem_size)
+        ]
+
+        # The remote memory starts anywhere after the local memory ends. We
+        # rely on the user to start and end this range.
+        self._remote_mem_ranges = [
+            self.get_remote_memory().get_mem_ports()[0][0]
```

Comment:
Why `[0][0]`? Should the interface to `get_mem_ports` be changed?

Reply:
So `get_mem_ports`' returns a tuple of address range and a port, which is directly used while setting up ruby caches. I did not want to create another function just to get the port or the range. I agree `[0][0]` does look hard-coded.